pydantic/utils.py

```diff
@@ -114,6 +137,84 @@ def almost_equal_floats(value_1: float, value_2: float, *, delta: float = 1e-8)
     return abs(value_1 - value_2) <= delta
 
 
+def copy_code(code: CodeType, **update: Any) -> CodeType:
+    if PY38:
+        return code.replace(**update)  # type: ignore
+    return CodeType(*[update.get(arg, getattr(code, arg)) for arg in CODE_ARGS])
+
+
+def generate_typed_init(
+    current_init: FunctionType,
+    fields: Dict[str, 'ModelField'],
+    defaults: Dict[str, Any],
+    annotations: Dict[str, Any],
+    compiled: bool,
+) -> FunctionType:
+    """
+    Generate typed signature for init
+    """
+    if not compiled:
+        # can perform this in python, however in cython this will cause "unknown opcode" on call
+        orig_init = current_init
+        orig_code = orig_init.__code__
+        new_init = FunctionType(
+            orig_code, orig_init.__globals__, closure=orig_init.__closure__, name=orig_init.__name__
+        )
+    else:
+        orig_init = getattr(current_init, '__origin_init__', current_init)
+        orig_code = orig_init.__code__
+
+        @wraps(orig_init)
+        def __init__(*args, **kwargs):  # type: ignore
+            orig_init(*args, **kwargs)
+
+        new_init = __init__  # type: ignore
+        new_init.__origin_init__ = orig_init  # type: ignore
+
+    orig_posonlycount = orig_code.co_posonlyargcount if PY38 else 0  # type: ignore
+    orig_allargscount = orig_code.co_argcount + orig_code.co_kwonlyargcount + orig_posonlycount
+    orig_argnames = orig_code.co_varnames[:orig_allargscount]
+    orig_othernames = orig_code.co_varnames[orig_allargscount:]
+
+    orig_argnames_set = set(orig_argnames)  # just for fast lookups
+    valid_fields_args = tuple([name for name in fields if name not in orig_argnames_set and name.isidentifier()])
+    del orig_argnames_set
+
+    fake_init = FunctionType(
+        copy_code(
+            orig_code,
+            co_argcount=orig_code.co_argcount,
+            co_kwonlyargcount=orig_code.co_kwonlyargcount + len(valid_fields_args),
+            co_names=orig_code.co_names,
+            co_varnames=orig_argnames + valid_fields_args + orig_othernames,
```

Comment:
Field order is wrong if we define a model like

```py
class Foo(BaseModel):
    a: int 
    b: str = 'xxx' 
    c: bool 
```
Which gives a signature of 
```
(__pydantic_self__, *, a: int, b: str = 'xxx', c: bool, **data: Any) -> None
```

Reply:
I don't see anything wrong here. Where's the wrong part?

Reply:
If you talking about the non-default parameter `c` after default `b`, there's nothing wrong since all of them are keyword-only:
```py
def f(*, a=1, b): ...
help(f)
'''
Help on function f in module __main__:

f(*, a=1, b)
'''
```

Reply:
good point, just me getting confused.